### PR TITLE
Port ComponentDesigner

### LINKED ISSUE #469: Required types and 'This method/object is not implemented by design.'
While trying to put our codebase on core3 rails, we faced unexpected exception that frustrated us.
 
The following types are strongly necessary for our application:
- System.Drawing.Design.ToolboxItem
- System.ComponentModel.Design.ComponentDesigner 
 
But we catch this exception: 
System.NotImplementedException: 'This method/object is not implemented by design.' 
 
Could you confirm that these types are in your plan and will be implemented in release v3.0?
 
We are also waiting for the following types from the System.Design lib:
- System.ComponentModel.Design.UndoEngine
- System.ComponentModel.Design.DesignSurface 
- System.ComponentModel.Design.InheritanceService 
- System.ComponentModel.Design.DesignerActionMethodItem
- System.ComponentModel.Design.DesignerActionPropertyItem
 
Can we expect to see all the missed types in the final release?

### PR DESCRIPTION
**Pull request by JuditRose**
Contributes to #221, #469

### CODE REVIEW

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerCommandSet.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.ComponentModel.Design;
+
+namespace System.Windows.Forms.Design.System.ComponentModel.Design
```
Comment (hughbe) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerCommandSet.cs]: is this right?

Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerCommandSet.cs]: good catch, no, this should be System.ComponentModel.Design

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs [hunk 2]
```diff
@@ -0,0 +1,607 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.Win32;
+using System.Diagnostics.CodeAnalysis;
+
+namespace System.Windows.Forms.Design.System.ComponentModel.Design
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs]: Is this the namespace of this file in Classic FX? Was it not System.ComponentModel.Design?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
+{
+#pragma warning disable 0436
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs]: If you just copy/pasted this, feel free to resolve as don't-know -- why are we disabling 0436?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs [hunk 4]
```diff
@@ -0,0 +1,607 @@
...
+    internal sealed class CodeMarkers
+    {
+        // Singleton access
+        public static readonly CodeMarkers Instance = new CodeMarkers();
+
+        static class NativeMethods
+        {
+#if Codemarkers_IncludeAppEnum
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs]: is Codemarkers_IncludeAppEnum defined for us? 

Also can we consolidate all these that are only in that to a single if / reorder this file?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs (hunk 2 above)
Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs]: fixed

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs (hunk 4 above)
Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs]: No, it's not defined for us, I removed them

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs (hunk 3 above)
Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs]: Good question, it was there originally. I removed it, it looks like working without it

File: src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj [hunk 5]
```diff
@@ -3,6 +3,7 @@
   <PropertyGroup>
     <TargetFramework>netcoreapp3.0</TargetFramework>
     <AssemblyName>System.Windows.Forms.Design</AssemblyName>    
+    <LangVersion>7.3</LangVersion>
```
Comment (JuditRose) [inline src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: Build failed: https://dev.azure.com/dnceng/public/_build/results?buildId=120915
For this reason: error CS8107: Feature 'default literal' is not available in C# 7.0. Please use language version 7.1 or greater.
According to [this GH issue](https://github.com/Microsoft/service-fabric/issues/142) comments adding lang version to csproj fixes it.

Comment (zsd4yr) [reply src/System.Windows.Forms.Design/src/System.Windows.Forms.Design.csproj]: We should probably make an effort to define this in a single place and import that into all the csproj's. I will create an issue for this.

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs [hunk 6]
```diff
@@ -0,0 +1,581 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.Win32;
+using System.Diagnostics.CodeAnalysis;
+
+namespace System.ComponentModel.Design
+{
+    internal sealed class CodeMarkers
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs]: Are there no tests for this file?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerCommandSet.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.ComponentModel.Design;
+
+namespace System.ComponentModel.Design
+{
+    public class DesignerCommandSet
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerCommandSet.cs]: Are there no tests for this file?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/InheritedPropertyDescriptor.cs [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+using System.Globalization;
+using System.Reflection;
+using System.Diagnostics.CodeAnalysis;
+
+namespace System.ComponentModel.Design
+{
+    /// <devdoc>
+    ///    <para>Describes and represents inherited properties in an inherited class.</para>
+    /// </devdoc>
+    internal class InheritedPropertyDescriptor : PropertyDescriptor
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/InheritedPropertyDescriptor.cs]: Are there no tests for this file? Could we add some?

Comment (JuditRose) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/InheritedPropertyDescriptor.cs]: Honestly my original plan was to add tests for the components that are on my list to port, and not for the dependencies that comes with them. We can add tests later for those.

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/InheritedPropertyDescriptor.cs [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+    {
+        private PropertyDescriptor propertyDescriptor;
+        private object _defaultValue;
+        private static object s_noDefault = new Object();
+        private bool _initShouldSerialize;
+
+        private object _originalValue;
+
+        /// <devdoc>
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/InheritedPropertyDescriptor.cs]: Consider replacing devdoc tags with summary and please remove the extra spaces on the next line

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs [hunk 10]
```diff
@@ -0,0 +1,581 @@
...
+            public static extern void TestDllPerfCodeMarkerString(int nTimerID, [MarshalAs(UnmanagedType.LPWStr, SizeParamIndex = 2)] string aUserParams, int cbParams);
+
+            [DllImport(ProductDllName, EntryPoint = "PerfCodeMarker")]
+            public static extern void ProductDllPerfCodeMarker(int nTimerID, [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2)] byte[] aUserParams, int cbParams);
+
+            [SuppressMessage("Microsoft.Performance", "CA1811:AvoidUncalledPrivateCode")]
+            [DllImport(ProductDllName, EntryPoint = "PerfCodeMarker")]
+            public static extern void ProductDllPerfCodeMarkerString(int nTimerID, [MarshalAs(UnmanagedType.LPWStr, SizeParamIndex = 2)] string aUserParams, int cbParams);
+
+            ///// global native method imports
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs]:
```suggestion
            // global native method imports
```

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs [hunk 11]
```diff
@@ -0,0 +1,581 @@
...
+using System.Runtime.InteropServices;
+using Microsoft.Win32;
+using System.Diagnostics.CodeAnalysis;
+
+namespace System.ComponentModel.Design
+{
+    internal sealed class CodeMarkers
+    {
+        // Singleton access
+        public static readonly CodeMarkers Instance = new CodeMarkers();
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs]: @ericstj  - do you know if we have plans to support CodeMarkers in Core? Who might know that?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs [hunk 12]
```diff
@@ -3,41 +3,64 @@
...
+using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Windows.Forms.Design;
 
 namespace System.ComponentModel.Design
 {
     /// <summary>
-    ///     <para>
     ///         The default designer for all components.
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs]:
```suggestion
    ///  The default designer for all components.
```

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs [hunk 13]
```diff
@@ -3,41 +3,64 @@
...
+        Hashtable _inheritedProps;
+        DesignerVerbCollection _verbs;
+        DesignerActionListCollection _actionLists;
+        ShadowPropertyCollection _shadowProperties;
+        bool _settingsKeyExplicitlySet = false;
+
+        private static readonly CodeMarkers s_codemarkers = CodeMarkers.Instance;
         /// <summary>
-        ///     <para>
         ///         Gets the design-time actionlists supported by the component associated with the designer.
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs]:
```suggestion
        ///  Gets the design-time actionlists supported by the component associated with the designer.
```

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs [hunk 14]
```diff
@@ -3,41 +3,64 @@
...
+                return _actionLists;
+            }
+        }
 
         /// <summary>
-        ///     <para>
-        ///         Retrieves a list of associated components.  These are components that should be incluced in a cut or copy
-        ///         operation on this component.
-        ///     </para>
+        ///         Retrieves a list of associated components. These are components that should be incluced in a cut or copy operation on this component.
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs]:
```suggestion
        ///  Retrieves a list of associated components. These are components that should be incluced in a cut or copy operation on this component.
```

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs (hunk 13 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs]: Could you please apply the same formatting across the file?

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/InheritedPropertyDescriptor.cs [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+            {
+                return !object.Equals(GetValue(component), _defaultValue);
+            }
+        }
+
+        private class ReadOnlyCollectionConverter : TypeConverter
+        {
+            public override object ConvertTo(ITypeDescriptorContext context, CultureInfo culture, object value, Type destinationType)
+            {
+
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/InheritedPropertyDescriptor.cs]: Please remove extra lines in this function

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs (hunk 11 above)
Comment (zsd4yr) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs]: @karelz might know something as well

Comment (karelz) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/CodeMarkers.cs]: Do you need it? Is it part of System.ComponentModel?

**Pull request merged**